From: Chad Monroe <chad@monroe.io>
Date: Tue, 18 Nov 2025 14:04:24 -0800
Subject: [PATCH] wifi: mt76: mt7996: add per-band token limit

Multi-radio devices use a single token pool for all bands. Add
per-band token limit to prevent one band from interfering with
another.

Add sysfs file to dump current token info for debugging.

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 mac80211.c        |    1 +
 mt76.h            |    9 ++++++++-
 mt76_connac_mac.c |    4 ++++
 mt7915/mac.c      |    3 ++-
 mt7921/pci_mac.c  |    4 +++-
 mt7925/pci_mac.c  |    4 +++-
 mt7996/debugfs.c  |   33 +++++++++++++++++++++++++++++++++
 mt7996/init.c     |    2 ++
 mt7996/mac.c      |    8 +++++++-
 mt7996/mt7996.h   |    1 +
 tx.c              |   21 ++++++++++++++++++---
 11 files changed, 82 insertions(+), 8 deletions(-)

--- a/mt76.h
+++ b/mt76.h
@@ -462,6 +462,9 @@ struct mt76_txwi_cache {
 	};
 
 	u8 qid;
+	u16 wcid;
+	u8 phy_idx;
+	unsigned long jiffies;
 };
 
 struct mt76_rx_tid {
@@ -921,6 +924,8 @@ struct mt76_phy {
 		bool al;
 		u8 pin;
 	} leds;
+
+	int tokens;
 };
 
 struct mt76_dev {
@@ -974,6 +979,7 @@ struct mt76_dev {
 	u16 token_count;
 	u16 token_start;
 	u16 token_size;
+	u16 token_threshold;
 
 	spinlock_t rx_token_lock;
 	struct idr rx_token;
@@ -2033,7 +2039,8 @@ static inline bool mt76_queue_is_npu_txf
 
 struct mt76_txwi_cache *
 mt76_token_release(struct mt76_dev *dev, int token, bool *wake);
-int mt76_token_consume(struct mt76_dev *dev, struct mt76_txwi_cache **ptxwi);
+int mt76_token_consume(struct mt76_dev *dev, struct mt76_txwi_cache **ptxwi,
+		       u8 phy_idx);
 void __mt76_set_tx_blocked(struct mt76_dev *dev, bool blocked);
 struct mt76_txwi_cache *mt76_rx_token_release(struct mt76_dev *dev, int token);
 int mt76_rx_token_consume(struct mt76_dev *dev, void *ptr,
--- a/mac80211.c
+++ b/mac80211.c
@@ -727,6 +727,7 @@ mt76_alloc_device(struct device *pdev, u
 	INIT_LIST_HEAD(&dev->txwi_cache);
 	INIT_LIST_HEAD(&dev->rxwi_cache);
 	dev->token_size = dev->drv->token_size;
+	dev->token_threshold = dev->token_size;
 	INIT_DELAYED_WORK(&dev->scan_work, mt76_scan_work);
 
 	for (i = 0; i < ARRAY_SIZE(dev->q_rx); i++)
--- a/mt76_connac_mac.c
+++ b/mt76_connac_mac.c
@@ -1216,8 +1216,12 @@ void mt76_connac2_tx_token_put(struct mt
 
 	spin_lock_bh(&dev->token_lock);
 	idr_for_each_entry(&dev->token, txwi, id) {
+		struct mt76_phy *phy = mt76_dev_phy(dev, txwi->phy_idx);
+
 		mt76_connac2_txwi_free(dev, txwi, NULL, NULL);
 		dev->token_count--;
+		if (phy->tokens > 0)
+			phy->tokens--;
 	}
 	spin_unlock_bh(&dev->token_lock);
 	idr_destroy(&dev->token);
--- a/tx.c
+++ b/tx.c
@@ -859,18 +859,28 @@ void __mt76_set_tx_blocked(struct mt76_d
 }
 EXPORT_SYMBOL_GPL(__mt76_set_tx_blocked);
 
-int mt76_token_consume(struct mt76_dev *dev, struct mt76_txwi_cache **ptxwi)
+int mt76_token_consume(struct mt76_dev *dev, struct mt76_txwi_cache **ptxwi,
+		       u8 phy_idx)
 {
-	int token;
+	struct mt76_phy *phy = mt76_dev_phy(dev, phy_idx);
+	int token = -EBUSY;
 
 	spin_lock_bh(&dev->token_lock);
 
+	if (dev->token_threshold &&
+	    phy->tokens >= dev->token_threshold)
+		goto out;
+
 	token = idr_alloc(&dev->token, *ptxwi, dev->token_start,
 			  dev->token_start + dev->token_size,
 			  GFP_ATOMIC);
-	if (token >= dev->token_start)
+	if (token >= dev->token_start) {
 		dev->token_count++;
 
+		(*ptxwi)->phy_idx = phy_idx;
+		phy->tokens++;
+	}
+
 #ifdef CONFIG_NET_MEDIATEK_SOC_WED
 	if (mtk_wed_device_active(&dev->mmio.wed) &&
 	    token >= dev->mmio.wed.wlan.token_start)
@@ -880,6 +890,7 @@ int mt76_token_consume(struct mt76_dev *
 	if (dev->token_count >= dev->token_size - MT76_TOKEN_FREE_THR)
 		__mt76_set_tx_blocked(dev, true);
 
+out:
 	spin_unlock_bh(&dev->token_lock);
 
 	return token;
@@ -908,12 +919,16 @@ struct mt76_txwi_cache *
 mt76_token_release(struct mt76_dev *dev, int token, bool *wake)
 {
 	struct mt76_txwi_cache *txwi;
+	struct mt76_phy *phy = NULL;
 
 	spin_lock_bh(&dev->token_lock);
 
 	txwi = idr_remove(&dev->token, token);
 	if (txwi) {
+		phy = mt76_dev_phy(dev, txwi->phy_idx);
 		dev->token_count--;
+		if (phy->tokens > 0)
+			phy->tokens--;
 
 #ifdef CONFIG_NET_MEDIATEK_SOC_WED
 		if (mtk_wed_device_active(&dev->mmio.wed) &&
--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -760,8 +760,9 @@ int mt7915_tx_prepare_skb(struct mt76_de
 
 	t = (struct mt76_txwi_cache *)(txwi + mdev->drv->txwi_size);
 	t->skb = tx_info->skb;
+	t->wcid = wcid->idx;
 
-	id = mt76_token_consume(mdev, &t);
+	id = mt76_token_consume(mdev, &t, wcid->phy_idx);
 	if (id < 0)
 		return id;
 
--- a/mt7921/pci_mac.c
+++ b/mt7921/pci_mac.c
@@ -26,8 +26,10 @@ int mt7921e_tx_prepare_skb(struct mt76_d
 
 	t = (struct mt76_txwi_cache *)(txwi + mdev->drv->txwi_size);
 	t->skb = tx_info->skb;
+	t->wcid = wcid->idx;
+	t->jiffies = jiffies;
 
-	id = mt76_token_consume(mdev, &t);
+	id = mt76_token_consume(mdev, &t, wcid->phy_idx);
 	if (id < 0)
 		return id;
 
--- a/mt7925/pci_mac.c
+++ b/mt7925/pci_mac.c
@@ -26,8 +26,10 @@ int mt7925e_tx_prepare_skb(struct mt76_d
 
 	t = (struct mt76_txwi_cache *)(txwi + mdev->drv->txwi_size);
 	t->skb = tx_info->skb;
+	t->wcid = wcid->idx;
+	t->jiffies = jiffies;
 
-	id = mt76_token_consume(mdev, &t);
+	id = mt76_token_consume(mdev, &t, wcid->phy_idx);
 	if (id < 0)
 		return id;
 
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -94,6 +94,7 @@
 #define MT7996_EXT_EEPROM_BLOCK_SIZE	1024
 #define MT7996_TOKEN_SIZE		16384
 #define MT7996_HW_TOKEN_SIZE		8192
+#define MT7996_PER_BAND_TOKEN_SIZE	4000
 
 #define MT7996_CFEND_RATE_DEFAULT	0x49	/* OFDM 24M */
 #define MT7996_CFEND_RATE_11B		0x03	/* 11B LP, 11M */
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -1749,6 +1749,8 @@ int mt7996_register_device(struct mt7996
 	if (ret)
 		return ret;
 
+	dev->mt76.token_threshold = MT7996_PER_BAND_TOKEN_SIZE;
+
 	mt7996_for_each_phy(dev, phy)
 		mt7996_thermal_init(phy);
 
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -1052,8 +1052,10 @@ int mt7996_tx_prepare_skb(struct mt76_de
 
 	t = (struct mt76_txwi_cache *)(txwi + mdev->drv->txwi_size);
 	t->skb = tx_info->skb;
+	t->wcid = wcid->idx;
+	t->jiffies = jiffies;
 
-	id = mt76_token_consume(mdev, &t);
+	id = mt76_token_consume(mdev, &t, wcid->phy_idx);
 	if (id < 0)
 		return id;
 
@@ -2210,8 +2212,12 @@ void mt7996_tx_token_put(struct mt7996_d
 
 	spin_lock_bh(&dev->mt76.token_lock);
 	idr_for_each_entry(&dev->mt76.token, txwi, id) {
+		struct mt76_phy *phy = mt76_dev_phy(&dev->mt76, txwi->phy_idx);
+
 		mt7996_txwi_free(dev, txwi, NULL, NULL, NULL);
 		dev->mt76.token_count--;
+		if (phy->tokens > 0)
+			phy->tokens--;
 	}
 	spin_unlock_bh(&dev->mt76.token_lock);
 	idr_destroy(&dev->mt76.token);
--- a/mt7996/debugfs.c
+++ b/mt7996/debugfs.c
@@ -820,6 +820,37 @@ mt7996_twt_stats(struct seq_file *s, voi
 	return 0;
 }
 
+static int mt7996_token_read(struct seq_file *s, void *data)
+{
+	struct mt7996_dev *dev = dev_get_drvdata(s->private);
+	struct mt76_dev *mdev = &dev->mt76;
+	int msdu_id, i;
+	struct mt76_txwi_cache *txwi;
+
+	seq_printf(s, "Token from host:\n");
+	spin_lock_bh(&mdev->token_lock);
+	idr_for_each_entry(&mdev->token, txwi, msdu_id) {
+		seq_printf(s, "%4d: wcid %4u band %u pending %u ms\n", msdu_id,
+			   txwi->wcid, txwi->phy_idx,
+			   jiffies_to_msecs(jiffies - txwi->jiffies));
+	}
+	seq_printf(s, "\n");
+	for (i = 0; i < __MT_MAX_BAND; i++) {
+		struct mt76_phy *phy = mt76_dev_phy(mdev, i);
+		u16 limit = mdev->token_threshold ?: mdev->token_size;
+
+		if (!mt7996_band_valid(dev, i))
+			continue;
+
+		seq_printf(s, "band%u consumed: %d, free: %d total: %d\n",
+			   i, phy->tokens,
+			   limit - phy->tokens, limit);
+	}
+	spin_unlock_bh(&mdev->token_lock);
+
+	return 0;
+}
+
 /* The index of RF registers use the generic regidx, combined with two parts:
  * WF selection [31:24] and offset [23:0].
  */
@@ -876,6 +907,8 @@ int mt7996_init_debugfs(struct mt7996_de
 			    &fops_implicit_txbf);
 	debugfs_create_devm_seqfile(dev->mt76.dev, "twt_stats", dir,
 				    mt7996_twt_stats);
+	debugfs_create_devm_seqfile(dev->mt76.dev, "token", dir,
+				    mt7996_token_read);
 	debugfs_create_file("rf_regval", 0600, dir, dev, &fops_rf_regval);
 
 	debugfs_create_u32("dfs_hw_pattern", 0400, dir, &dev->hw_pattern);
